File: src/ts/src/strategy/Sorter.test.ts
// Import the Sorter class
import { Sorter } from "./Sorter"; // Correct for ES module
type SorterType = InstanceType<typeof Sorter>;

describe("Sorter", () => {
  let sorter: SorterType;

  test("should sort data using bubbleSort", () => {
    const data = [5, 2, 9, 1, 5, 6];
    const sortedData = Sorter.sort("bubbleSort", data);
    expect(sortedData).toEqual([1, 2, 5, 5, 6, 9]);
  });

  test("should sort data using mergeSort", () => {
    const data = [38, 27, 43, 3, 9, 82, 10];
    const sortedData = Sorter.sort("mergeSort", data);
    expect(sortedData).toEqual([3, 9, 10, 27, 38, 43, 82]);
  });

  test("should sort data using quickSort", () => {
    const data = [10, 7, 8, 9, 1, 5];
    const sortedData = Sorter.sort("quickSort", data);
    expect(sortedData).toEqual([1, 5, 7, 8, 9, 10]);
  });

  test("should throw an error for an unsupported sorting algorithm", () => {
    const data = [3, 1, 2];
    expect(() => Sorter.sort("unknownSort", data)).toThrow(
      "Unknown algorithm: unknownSort"
    );
  });
});

File: src/java/design-patterns/src/test/java/strategy/solution/SorterTest.java
package strategy.solution;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;
import strategy.solution.Sorter;

public class SorterTest {

  @Test
  public void testBubbleSortUsingSorter() {
    Sorter sorter = new Sorter();
    int[] data = {4, 2, 9, 7, 3};
    int[] sortedData = sorter.sort("bubbleSort", data);
    assertArrayEquals(new int[]{2, 3, 4, 7, 9}, sortedData);
  }

  @Test
  public void testQuickSortUsingSorter() {
    Sorter sorter = new Sorter();
    int[] data = {3, 6, 8, 10, 1, 2, 1};
    int[] sortedData = sorter.sort("quickSort", data);
    assertArrayEquals(new int[]{1, 1, 2, 3, 6, 8, 10}, sortedData);
  }

  @Test
  public void testMergeSortUsingSorter() {
    Sorter sorter = new Sorter();
    int[] data = {38, 27, 43, 3, 9, 82, 10};
    int[] sortedData = sorter.sort("mergeSort", data);
    assertArrayEquals(new int[]{3, 9, 10, 27, 38, 43, 82}, sortedData);
  }

  @Test
  public void testUnknownAlgorithmThrowsException() {
    Sorter sorter = new Sorter();
    int[] data = {1, 2, 3};
    Exception exception = assertThrows(IllegalArgumentException.class, () -> {
      sorter.sort("unknownSort", data);
    });
    assertEquals("Unknown sorting algorithm: unknownSort", exception.getMessage());
  }
}
